File: leetcode/pacific-atlantic-water-flow.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
	public List<List<Integer>> pacificAtlantic(int[][] heights) {
	   List<List<Integer>> coords = new ArrayList<List<Integer>>();
	   boolean[][] visited = new boolean[heights.length][heights[0].length];
	   boolean[] flow = new boolean[2];
		
	   for(int i = 0; i < heights.length; i++) {
		   for(int j = 0; j < heights[0].length; j++) {
			   flow[0] = false;
			   flow[1] = false;
			   dfs(heights,i, j, flow, visited);
			   if(flow[0] && flow[1]) coords.add(Arrays.asList(i, j));
		   }
	   }
	   return coords;
   }
   
   private void dfs(int[][] heights, int row, int col, boolean[] flow, boolean[][] visited) {
	   if((!flow[0] || !flow[1]) && (row < heights.length && col < heights[0].length && row >= 0 && col >=0 && !visited[row][col])){
	   
		   visited[row][col] = true;
		   
		   if(row == 0 || col == 0) flow[0] = true; //reached pacific ocean
		   if(row == heights.length-1 || col == heights[0].length-1) flow[1] = true; //reached atlantic ocean

		   if(row+1 < heights.length) {
			   if(heights[row][col] >= heights[row+1][col])
				   dfs(heights, row+1, col, flow, visited);
		   }
		   if(col+1 < heights[0].length) {
			   if(heights[row][col+1] <= heights[row][col])
				   dfs(heights, row, col+1, flow, visited);
		   }
		   if(row-1 >= 0) {
			   if(heights[row-1][col] <= heights[row][col])
				   dfs(heights, row-1, col, flow, visited);
		   }
		   if(col-1 >= 0) {
			   if(heights[row][col-1] <= heights[row][col])
				   dfs(heights, row, col-1, flow, visited);
		   }
		   
		   // Reset visited matrix
		   visited[row][col] = false;
	   }
   }
}

File: leetcode/pacific_atlantic_water_flow.py
# https://leetcode.com/problems/pacific-atlantic-water-flow


class Solution:
    def pacificAtlantic(self, heights: list[list[int]]) -> list[list[int]]:
        pacific: list[int] = []
        atlantic: list[int] = []

        for i in range(len(heights)):
            pacific.append((i, 0))
            atlantic.append((i, len(heights[0]) - 1))

        for j in range(len(heights[0])):
            pacific.append((0, j))
            atlantic.append((len(heights) - 1, j))

        reach_pacific: set[int] = self.bfs(pacific, heights)
        reach_atlantic: set[int] = self.bfs(atlantic, heights)

        return list(reach_pacific.intersection(reach_atlantic))

    def bfs(self, queue: list[tuple], heights: list[list[int]]) -> set:
        seen = set()

        while queue:
            x, y = queue.pop(0)
            seen.add((x, y))

            for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                if (
                    (0 <= x + dx < len(heights))
                    and (0 <= y + dy < len(heights[0]))
                    and (heights[x + dx][y + dy] >= heights[x][y])
                    and (x + dx, y + dy) not in seen
                ):
                    queue.append((x + dx, y + dy))

        return seen
